File: tests/home_exhibition_layout_test.py
from pathlib import Path
import re


ROOT = Path(__file__).resolve().parents[1]
HTML = (ROOT / "site/index.html").read_text()
CSS = (ROOT / "site/styles.css").read_text()
SCRIPT = (ROOT / "site/script.js").read_text()
EXHIBITION_SCRIPT = (ROOT / "site/exhibition/exhibition-test.js").read_text()


def test_home_exhibition_has_six_cd_case_cards():
    cards = re.findall(r'class="art-card\b[^>]*"', HTML)
    assert len(cards) == 6
    assert "grid-template-columns: repeat(3" in CSS
    assert "art-card__spine" in HTML
    assert "art-card__cover" in HTML
    assert "aspect-ratio: 1.08 / 1" in CSS
    assert ".slice(0, 6)" in SCRIPT


def test_each_cd_row_sits_on_its_shelf_without_overflow():
    gallery_rule = re.search(r"\.gallery-row \{([^}]+)\}", CSS).group(1)
    assert "height: auto" in gallery_rule
    assert "grid-template-rows" not in gallery_rule
    assert "row-gap: 30px" in gallery_rule
    assert "top: calc(50% - 15px)" in CSS


def test_home_cd_links_open_the_single_assigned_role():
    assert "work.roles?.[0]" in SCRIPT
    assert 'view=roles&role=' in SCRIPT


def test_admin_can_choose_and_order_six_home_images():
    assert "work.roles.length === 1" in EXHIBITION_SCRIPT
    assert ".slice(0, 6)" in EXHIBITION_SCRIPT
    assert "selected.length < 6" in EXHIBITION_SCRIPT
    assert "按 01 至 06" in EXHIBITION_SCRIPT


def test_exhibition_accepts_a_role_deep_link():
    assert "new URLSearchParams(window.location.search)" in EXHIBITION_SCRIPT
    assert 'requestedView === "roles"' in EXHIBITION_SCRIPT


if __name__ == "__main__":
    test_home_exhibition_has_six_cd_case_cards()
    test_each_cd_row_sits_on_its_shelf_without_overflow()
    test_home_cd_links_open_the_single_assigned_role()
    test_admin_can_choose_and_order_six_home_images()
    test_exhibition_accepts_a_role_deep_link()
    print("home exhibition layout test: PASS")

File: tests/home-exhibition-layout.test.mjs
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import test from "node:test";

const html = readFileSync(new URL("../site/index.html", import.meta.url), "utf8");
const css = readFileSync(new URL("../site/styles.css", import.meta.url), "utf8");

test("the homepage exhibition keeps six covers in two three-card rows", () => {
  assert.equal((html.match(/class="art-card\b/g) || []).length, 6);
  assert.equal((html.match(/class="gallery-level\b/g) || []).length, 2);
  assert.equal((html.match(/class="gallery-shelf"/g) || []).length, 2);
  assert.match(css, /\.gallery-level__cards\s*\{[^}]*grid-template-columns:\s*repeat\(3,/s);
});

test("each shelf supports its row instead of crossing through it", () => {
  const levelRule = css.match(/\.gallery-level\s*\{([^}]*)\}/s)?.[1] || "";
  const shelfRule = css.match(/\.gallery-shelf\s*\{([^}]*)\}/s)?.[1] || "";

  assert.match(levelRule, /padding-bottom:/);
  assert.match(shelfRule, /bottom:/);
  assert.doesNotMatch(css, /top:\s*calc\(50%/);
});
